server/src/main/java/com/cloud/usage/UsageServiceImpl.java

```diff
@@ -248,7 +248,17 @@ public boolean generateUsageRecords(GenerateUsageRecordsCmd cmd) {
             sc.addAnd("domainId", SearchCriteria.Op.IN, domainIds.toArray());
         }
 
-        if (domainId != null) {
+        if (cmd.isRecursive() && domainId != null){
```

Comment:
Look at for reference https://github.com/apache/cloudstack/blob/master/server/src/main/java/com/cloud/api/query/QueryManagerImpl.java#L872

Expected test if an account is allowed or disallowed to get usage records of other domains (with/without recursive flag), and allowed to recursively get all usage records for a permitted account/domain.

Reply:
While testing I found that users cannot call this API, but domain admins can. I have added a check, if an account id is specified, if it belongs to any of the domains for the calling domain admin.